lib/index.js

```diff
@@ -31,12 +66,12 @@ export default class Pageres {
 	 * @api public
 	 */
 
-	src(url, sizes, options) {
+	// src(url?: string, sizes?: Array<string>, options?: Options): Pageres | Array<Src> {
+	src(url, sizes, options): Pageres | Array<Src> { // TODO: how can I make all the arguments optional or all required? the commented out line doesn't work as it complains it can also be `undefined`. Can I define a separate type definiton for each scenario?
```

Comment:
`Value` is a polymorphic or "generic" type, basically it's a way of following a value around a function or class. 

Although I didn't realize that src was an array of sources so you can't do that. You should do this:


```js
type Options = {
  delay?: number;
  // ...
};

type Src = {
  url: string;
  sizes: Array<string>;
  options: Options;
};

type SrcFn<DestValue> =
  & ((_: void, _: void, _: void) => Array<Src>)
  & ((url: string, sizes: Array<string>, options: Options) => Pageres<DestValue>); // return a Pageres with the same dest, you can't pass "this" because we're using a class property type for the intersection

type DestFn<DestValue> =
  & ((_: void) => DestValue)
  & ((dir: DestValue) => Pageres<DestValue>); // return a Pageres with the same dest (same as before)

class Pageres<DestValue: string> {
  _src: Array<Src>;
  _dest: DestValue;

  src: SrcFn<DestValue>;
  src(url: string, sizes: Array<string>, options: Options) {
    if (!url) {
      return this._src;
    } else {
      this._src.push({ url, sizes, options });
      return this;
    }
  }

  dest: DestFn<DestValue>;
  dest(dir: DestValue) {
    if (dir === undefined) {
      return this._dest;
    } else {
      this._dest = dir;
      return this;
    }
  }
}

let pageres = new Pageres();
pageres.src("hi", [], {});
let src: Array<Src> = pageres.src();
let dest: string = pageres.dest("foo").dest();
```